external/llvm-project/mlir/include/mlir/ExecutionEngine/SystemDetect.h

```diff
@@ -0,0 +1,54 @@
+//===- SystemDetect.h - ExecutionEngine System Detect ----------*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides a system detection interface with default CPU detection
+// for the execution engine.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_EXECUTIONENGINE_SYSTEMDETECT_H_
+#define MLIR_EXECUTIONENGINE_SYSTEMDETECT_H_
+
+#include <string>
+#include <unordered_map>
+
+namespace mlir {
+
+/// SystemDetect finds devices on the current system.
+///
+
+struct SystemDevice {
+  enum class Type : int32_t { ECPU, EGPU, ENPU, EALT };
+
+  Type type;
+  uint32_t count;
+  std::string chip;
+  std::unordered_map<std::string, std::string> properties;
```

Comment:
For example, this is almost certainly better as a `StringMap`

Reply:
StringMap does not own the keys (StringRef) and there are cases where property keys are not constant.

Reply:
StringMap does own the keys - it just takes `StringRef`s in the API to allow passing keys by reference